src/core_codemods/use_walrus_if.py

```diff
@@ -27,52 +34,85 @@ class UseWalrusIf(SemgrepCodemod):
         }
     ]
 
-    @classmethod
-    def rule(cls):
-        return """
-        rules:
-          - patterns:
-            - pattern: |
-                $ASSIGN
-                if $COND:
-                  $BODY
-            - metavariable-pattern:
-                metavariable: $ASSIGN
-                patterns:
-                  - pattern: $VAR = $RHS
-                  - metavariable-pattern:
-                      metavariable: $COND
-                      patterns:
-                        - pattern: $VAR
-                  - metavariable-pattern:
-                      metavariable: $BODY
-                      pattern: $VAR
-            - focus-metavariable: $ASSIGN
-        """
-
-    _modify_next_if: List[Tuple[CodeRange, cst.Assign]]
-    _if_stack: List[Optional[Tuple[CodeRange, cst.Assign]]]
+    _modify_next_if: List[Tuple[CodeRange, cst.NamedExpr]]
+    _if_stack: List[Optional[Tuple[CodeRange, cst.NamedExpr]]]
+    assigns: dict[cst.Assign, cst.NamedExpr]
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._modify_next_if = []
         self._if_stack = []
+        self.assigns = {}
+
+    def _build_named_expr(self, target, value, parens=True):
+        return cst.NamedExpr(
+            target=target,
+            value=value,
+            lpar=[cst.LeftParen()] if parens else [],
+            rpar=[cst.RightParen()] if parens else [],
+        )
 
-    def visit_If(self, node):
+    def on_visit(self, node: cst.CSTNode) -> Optional[bool]:
```

Comment:
It seems the purpose of this method is to identify two consecutive statements each that satisfies a predicate.

The way it is, it is a bit hard to understand right now. My suggestion is to break this into a few methods: (1) `handle_block`, (2) `is_name_assign() -> Optional[cst.Assign]`, (3) `is_named_comparison_if -> Optional[cst.If]`. As always, suggestions are optional.

Functions (2) and (3) are straightforward and just test the desired properties.  For (1), there are three nodes that may contain a sequence of statements (actually 4 since `cst.SimpleStatementLine` may contain more than one statement, but we should ignore that for now): `cst.Module`, `cst.IndentedBlock` and `cst.SimpleStatementSuite`. They all have the `body` attribute which contains the statements in order. So you can use `visit_Module`, `visit_SimpleStatementSuite` and `visit_IndentedBlock` all which will simply call `handle_block`.

Thus the implemenation of `handle_block` would look like:
```
for s1, s2 in pairwise(original_node.body):
  assign = is_name_assign(s1)
  if_stmt = is_named_comparison_if(s2)
  if assign and if_stmt:
     <do what you need>
```

Reply:
Good suggestion. This method is a bit unwieldy right now. The `match` statements are really useful but they add a lot of indentation.